src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1467,8 +1467,15 @@ private TypeDeclarationSyntax ParseClassOrStructOrInterfaceDeclaration(SyntaxLis
             {
                 if (this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword)
                 {
+                    var saveTerm2 = _termState;
+                    if (keyword.Kind == SyntaxKind.RecordKeyword)
+                    {
+                        _termState |= TerminatorState.IsEndOfMethodSignature;
```

Comment:
>IsEndOfMethodSignature [](start = 54, length = 22)

1) I think we want to add special member for end of record header, perhaps with the same behavior, but specific for records.
2) I think we want to move adjustment for ```_termState``` to the beginning of this method. So that ```ParseTypeParameterList``` and ```ParseBaseList``` calls could take advantage of it as well. We can remove it right before ```if (!(keyword.Kind == SyntaxKind.RecordKeyword) || CurrentToken.Kind != SyntaxKind.SemicolonToken)```.
3) I think we want to remove explicit checks for semicolon from ```ParseBaseList``` and ```SkipBadBaseListTokens```. I believe those were added there specifically to support records and we should rely on existing infrastructure instead. That will also prevent us from affecting non-record declarations (non-critical, I believe, but if it comes for free, I would take this).